### PR TITLE
Optimized handling case-insensitive mappings.

### PR DESCRIPTION
**Pull request by illia-v**
Elements yielded by `_destruct_iterable_mapping_values` are always unpacked. Since unpacking can be done with any iterable, there is no need to convert elements to tuples if they are iterables.

Also, such elements can be used directly in for loops, creating a dictionary of them is excessive.

### CODE REVIEW

**Comment by auvipy** [comment]
@pope1ni

File: tests/utils_tests/test_datastructures.py [hunk 1]
```diff
@@ -275,6 +275,12 @@ def test_create_with_invalid_key(self):
         with self.assertRaisesMessage(ValueError, msg):
             CaseInsensitiveMapping([(1, '2')])
 
+    def test_create_with_not_mappings(self):
+        data = {'Accept': 'application/json'}
+        for iterable in (tuple, list, ImmutableList):
+            mapping = CaseInsensitiveMapping(map(iterable, data.items()))
+            self.assertEqual(mapping, data)
```
Comment (pope1ni) [inline tests/utils_tests/test_datastructures.py:282]: This test passes before the changes so please split it out into a separate commit.

File: tests/utils_tests/test_datastructures.py [hunk 2]
```diff
@@ -275,6 +275,12 @@ def test_create_with_invalid_key(self):
         with self.assertRaisesMessage(ValueError, msg):
             CaseInsensitiveMapping([(1, '2')])
 
+    def test_create_with_not_mappings(self):
```
Comment (pope1ni) [inline tests/utils_tests/test_datastructures.py:278]:
```suggestion
    def test_create_with_non_mapping(self):
```

File: tests/utils_tests/test_datastructures.py [hunk 3]
```diff
@@ -275,6 +275,12 @@ def test_create_with_invalid_key(self):
...
+        data = {'Accept': 'application/json'}
+        for iterable in (tuple, list, ImmutableList):
```
Comment (pope1ni) [inline tests/utils_tests/test_datastructures.py:280]: Please use `with self.subTest(...):` in the loop.

File: django/utils/datastructures.py [hunk 4]
```diff
@@ -286,7 +286,7 @@ def _destruct_iterable_mapping_values(data):
             )
         if not isinstance(elem[0], str):
             raise ValueError('Element key %r invalid, only strings are allowed' % elem[0])
-        yield tuple(elem)
+        yield elem if isinstance(elem, Iterable) else tuple(elem)
```
Comment (pope1ni) [inline django/utils/datastructures.py:289]: No tests fail if I remove this change. I presume it was to avoid calling `tuple()` unnecessarily?

For that matter, we can just change to the following as we're always unpacking into key-value pairs:

```suggestion
        yield elem
```

I don't think we need any guarantees that this will return a `tuple`. Loosely it only needs return an iterable of two elements with the first being a string.

File: django/utils/datastructures.py [hunk 5]
```diff
@@ -308,9 +308,11 @@ class CaseInsensitiveMapping(Mapping):
     """
 
     def __init__(self, data):
-        if not isinstance(data, Mapping):
-            data = {k: v for k, v in _destruct_iterable_mapping_values(data)}
-        self._store = {k.lower(): (k, v) for k, v in data.items()}
+        if isinstance(data, Mapping):
+            data_items = data.items()
+        else:
+            data_items = _destruct_iterable_mapping_values(data)
```
Comment (pope1ni) [inline django/utils/datastructures.py:314]: There are only two uses of `_destruct_iterable_mapping_values()` and we use this same pattern exactly.

I think that you could push the `isinstance(..., Mapping)` check into that function.

Comment (pope1ni) [reply django/utils/datastructures.py:314]: As a separate commit, I think that it is also probably worth moving that global function to be a `@staticmethod` on the `CaseInsensitiveMapping` class. It is closely associated with this class and would avoid the need to import it separately in `django.http.response`.

**Review by pope1ni — CHANGES_REQUESTED** [review]

**Comment by felixxm** [comment]
@illia-v Do you have time to keep working on this?

**Comment by illia-v** [comment]
> @illia-v Do you have time to keep working on this?

Sorry for the delay, I'll try to work on this next week

File: django/utils/datastructures.py (hunk 5 above)
Comment (illia-v) [reply django/utils/datastructures.py:321]: > There are only two uses of `_destruct_iterable_mapping_values()` and we use this same pattern exactly.
> 
> I think that you could push the `isinstance(..., Mapping)` check into that function.

`_destruct_iterable_mapping_values` is a generator function, moving the condition to it will result in adding an additional `for`-loop to iterate over mappings

File: django/utils/datastructures.py (hunk 4 above)
Comment (illia-v) [reply django/utils/datastructures.py]: Thanks. I dropped calling `tuple`

File: tests/utils_tests/test_datastructures.py (hunk 1 above)
Comment (illia-v) [reply tests/utils_tests/test_datastructures.py]: I removed the test from the changeset

File: tests/utils_tests/test_datastructures.py (hunk 2 above)
Comment (illia-v) [reply tests/utils_tests/test_datastructures.py]: I removed the test from the changeset

File: tests/utils_tests/test_datastructures.py (hunk 3 above)
Comment (illia-v) [reply tests/utils_tests/test_datastructures.py]: I removed the test from the changeset

File: django/utils/datastructures.py [hunk 6]
```diff
@@ -285,6 +285,9 @@ def __getitem__(self, key):
 
 
 def _destruct_iterable_mapping_values(data):
```
Comment (ngnpope) [inline django/utils/datastructures.py:287]: So the last thing I'd suggest is moving this to be a `@staticmethod` on `CaseInsensitiveMapping`, and rename it to something like `_unpack_values()`.

This means that we don't need to import the separate private function into `django.http.response` and can call `self.unpack_values(data)`.

File: django/http/response.py [hunk 7]
```diff
@@ -32,10 +31,9 @@ def __init__(self, data):
         Populate the initial data using __setitem__ to ensure values are
         correctly encoded.
         """
-        if not isinstance(data, Mapping):
-            data = {k: v for k, v in _destruct_iterable_mapping_values(data)}
+        data_items = _destruct_iterable_mapping_values(data)
         self._store = {}
-        for header, value in data.items():
+        for header, value in data_items:
```
Comment (ngnpope) [inline django/http/response.py:36]: Then this can just be:
```python
        for header, value in self._unpack_values(data):
```

**Review by ngnpope — CHANGES_REQUESTED** [review]

File: django/utils/datastructures.py (hunk 6 above)
Comment (illia-v) [reply django/utils/datastructures.py]: Done. I named it `_unpack_items` though :)

**Comment by felixxm** [comment]
Thanks both :+1:
